### PR TITLE
Fix eventual consistency issues in Shared Directory regarding Reconnection

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Fix eventual consistency issues in Shared Directory regarding after disconnection and reconnection. Also enable the reconnect fuzz tests.

1.) On reconnection, it could happen that we don't need to submit pending ops for a sub directory because that subdirectory could have been deleted and created again in the meantime, so if we don't have pending record then don't resubmit the op.
2.) When we process a op to delete a subdirectory, but we have pending create for that subdirectory and maybe pending create/set/clear for child subdirectories, then just delete the ones for which pending create is not pending as they will be created eventually.
3.) Don't need to process create in case it is for a different instance of subdirectory.

### CODE REVIEW

**Review by Abe27342 — COMMENTED** [review]
could you add a regression test for the issue? and brief description to the PR of the technical problem here :)

**Comment by jatgarg** [comment]
> could you add a regression test for the issue? and brief description to the PR of the technical problem here :)

Doesn't fuzz testing cover different cases related to reconnection, add new clients etc?

**Comment by Abe27342** [comment]
> > could you add a regression test for the issue? and brief description to the PR of the technical problem here :)
> 
> Doesn't fuzz testing cover different cases related to reconnection, add new clients etc? I tried this with different values of reconnect probability.

it does, but taking the time to write a more targeted test is useful for more quickly diagnosing problems in the future--a fuzz test failing only tells you "there's something wrong" whereas a handwritten one will point you much closer to the actual problem without needing to debug.

File: packages/dds/map/src/directory.ts [hunk 1]
```diff
@@ -1106,10 +1103,16 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
...
 	/**
-	 * Subdirectories that have been deleted locally but not yet ack'd from the server. This maintains the count
+	 * Subdirectories that have been deleted locally but not yet ack'd from the server. This maintains the record
 	 * of delete op that are pending or yet to be acked from server.
 	 */
-	private readonly pendingDeleteSubDirectoriesCount: Map<string, number> = new Map();
+	private readonly pendingDeleteSubDirectoriesTracker: Set<string> = new Set();
+
+	/**
+	 * Subdirectories that have been created locally but not yet ack'd from the server. This maintains the record
```
Comment (Abe27342) [inline packages/dds/map/src/directory.ts:1112]: could you clarify here and in the doc comments for `pendingSubDirectories` / `pendingKeys` how the two fields relate? It's not clear why we need both. It's probably also worth documenting that the `number[]` types in `pendingSubDirectories` and `pendingKeys` are lists of pendingMessageIds at which that key was modified.

**Review by Abe27342 — COMMENTED** [review]
still working through the logic changes, have a few pieces of high-level feedback from initial review

File: packages/dds/map/src/directory.ts (hunk 1 above)
Comment (jatgarg) [reply packages/dds/map/src/directory.ts:1119]: done

File: packages/dds/map/src/directory.ts [hunk 2]
```diff
@@ -1096,20 +1093,33 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
...
 	private readonly pendingSubDirectories: Map<string, number[]> = new Map();
 
 	/**
-	 * Subdirectories that have been deleted locally but not yet ack'd from the server. This maintains the count
-	 * of delete op that are pending or yet to be acked from server.
+	 * Subdirectories that have been deleted locally but not yet ack'd from the server. This maintains the record
+	 * of delete op that are pending or yet to be acked from server. This is maintained just to track the locally
+	 * deleted sub directory as pendingSubDirectories does not track that specifically.
+	 */
+	private readonly pendingDeleteSubDirectoriesTracker: Set<string> = new Set();
```
Comment (Abe27342) [inline packages/dds/map/src/directory.ts:1116]: I don't follow how logic in this area avoids the need to count number of times the directory may have been deleted/recreated while pending. Ex: client A disconnects and while disconnected, creates `/dir` and deletes `/dir` 3 times, then reconnects. I think there are still bugs in this area: if I check out this branch and modify fuzz test config to lean heavily towards creation/deletion of the same directory (by for example limiting the directory name pool to have 2 options and only generating subdirectory create/delete ops), I see lots of failures with json that resembles the above kind of thing. See [this commit](https://github.com/Abe27342/FluidFramework/commit/ff212cab4fafae2061795dd09a4faf07f529f494).

**Review by Abe27342 — COMMENTED** [review]
sorry for the delayed review--got around to confirming my hypothesis. The general approach here looks like a step in the right direction, I think the pending state tracking just doesn't quite work as-is.

File: packages/dds/map/src/directory.ts (hunk 2 above)
Comment (jatgarg) [reply packages/dds/map/src/directory.ts]: resolved.

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts [hunk 3]
```diff
@@ -294,61 +313,45 @@ function makeReducer(loggingInfo?: LoggingInfo): AsyncReducer<Operation, FuzzTes
...
-		assertEventualConsistencyCore(subDirectory1, subDirectory2);
-	}
-}
+	createDDSFuzzSuite(model, {
+		validationStrategy: { type: "fixedInterval", interval: defaultOptions.validateInterval },
+		reconnectProbability: 0.15,
+		numberOfClients: 3,
+		clientJoinOptions: {
+			// Note: if tests are slow, we may want to tune this down. This mimics behavior before this suite
+			// was refactored to use the DDS fuzz harness.
```
Comment (Abe27342) [inline packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:341]:
```suggestion
```

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**